src/Components/Server/src/Circuits/CircuitRegistry.cs

```diff
@@ -0,0 +1,243 @@
+        }
+
+        protected virtual bool DisconnectCore(CircuitHost circuitHost, string connectionId)
+        {
+            if (!ConnectedCircuits.TryGetValue(circuitHost.CircuitId, out circuitHost))
+            {
+                // Guard: The circuit might already have been marked as inactive.
+                return false;
+            }
+
+            if (!string.Equals(circuitHost.Client.ConnectionId, connectionId, StringComparison.Ordinal))
+            {
+                // The circuit is associated with a different connection. One way this could happen is when
+                // the client reconnects with a new connection before the OnDisconnect for the older
+                // connection is executed. Do nothing
+                return false;
+            }
+
+            var result = ConnectedCircuits.TryRemove(circuitHost.CircuitId, out circuitHost);
+            Debug.Assert(result, "This operation operates inside of a lock. We expect the previously inspected value to be still here.");
+
+            circuitHost.Client.SetDisconnected();
+            var entryOptions = new MemoryCacheEntryOptions
+            {
+                AbsoluteExpiration = DateTimeOffset.UtcNow.Add(_options.DisconnectedCircuitRetentionPeriod),
+                Size = 1,
+                PostEvictionCallbacks = { _postEvictionCallback },
+            };
+
+            DisconnectedCircuits.Set(circuitHost.CircuitId, circuitHost, entryOptions);
+            return true;
+        }
+
+        public virtual async Task<CircuitHost> ConnectAsync(string circuitId, IClientProxy clientProxy, string connectionId, CancellationToken cancellationToken)
+        {
+            CircuitHost circuitHost = null;
+
+            lock (CircuitRegistryLock)
+            {
+                // Transition the host from disconnected to connected if it's available. In this critical section, we return
+                // an existing host if it's currently considered connected or transition a disconnected host to connected.
+                // Transfering also wires up the client to the new set.
+                circuitHost = ConnectCore(circuitId, clientProxy, connectionId);
+            }
+
+            if (circuitHost == null)
+            {
+                return null;
+            }
+
+            // CircuitHandler events are invoked outside the critical section. This may result in simultaneous
+            // execution of a disconnect and connect events, however we make no guarantees of the order in which they execute.
+            // If we transfered a circuit that was considered active, we will invoke OnConnectionUpAsync without having
+            // a corresponding OnConnectionDownAsync.
+            await circuitHost.OnConnectionUpAsync(cancellationToken);
+
+            // If we acccumulated any renders during the disconnect, fire it off in the background. We don't need to wait for it to complete.
+            _ = circuitHost.Renderer.DispatchBufferedRenderAsync();
```

Comment:
Why don't we `await` this? Could you clarify why it should be fire-and-forget?

Reply:
+1